ltm/inserted_context_agent.py

```diff
@@ -0,0 +1,577 @@
+            interaction = self.messages_from_memory(m)
+            if interaction not in full_interactions:
+                full_interactions.append(interaction)
+
+        for m in full_interactions:
+            if "trivia" in m[0].content:
+                colour_print("YELLOW", f"<*** trivia ***>")
+            else:
+                colour_print("YELLOW", f"{m[0].content}")
+
+        # Add the previous messages
+        final_idx = self.session.message_count - 1
+        while previous_interactions > 0 and final_idx > 0:
+
+            # Agent reply
+            context.insert(1, self.session.by_index(final_idx).as_llm_dict())
+            # User message
+            context.insert(1, self.session.by_index(final_idx-1).as_llm_dict())
+
+            final_idx -= 2
+            previous_interactions -= 1
+
+        # Add in memories up to the max prompt size
+        current_size = token_counter(self.model, messages=context)
+        shown_mems = 0
+        target_size = max_prompt_size - self.max_message_size
+
+        for interaction in full_interactions[::-1]:
+            user_message, assistant_message = interaction
+            future_size = current_size + token_counter(self.model, messages=[user_message.as_llm_dict(), assistant_message.as_llm_dict()])
+
+            # If this message is going to be too big, then skip it
+            if future_size > target_size or shown_mems == 100:
+                continue
+
+            # Add the interaction and count the tokens
+            if user_message not in context:
+                context.insert(1, assistant_message.as_llm_dict())
+                context.insert(1, user_message.as_llm_dict())
+                shown_mems += 1
+
+                current_size = future_size
+
+        print(f"current context size: {current_size}")
+
+        return context
+
+    def llm_memory_filter(self, memories, queries, keywords, cost_cb):
+
+        situation_prompt = """You are a part of an agent. Another part of the agent is currently searching for memories using the statements below.
+Based on these statements, describe what is currently happening external to the agent in general terms:
+{queries}  
+"""
+
+        prompt = """Here are a number of interactions, each is given a number:
+{passages}         
+*********
+
+Each of these interactions might be related to the general situation below. Your task is to judge if these interaction have any relation to the general situation.
+Filter out interactions that very clearly do not have any relation. But keep in interactions that have any kind of relationship to the situation such as in: topic, characters, locations, setting, etc.
+
+SITUATION:
+{situation}
+
+Express your answer in this JSON: 
+[
+    {{
+        "number": int  // The number of the interaction.
+        "justification": string  // Why the interaction is or is not related to the situation.
+        "related": bool // Whether the interaction is related to the situation.
```

Comment:
I think this is nice, and we should apply the idea more often:
1. Force the LLM into a decision.
2. Make it write a justification to have it think about it.
3. Allow it to retract.